### DISCUSSION TITLE
How to group/aggregate by time interval?

### DISCUSSION THREAD

**Discussion by jklingen** [grafana/grafana-infinity-datasource#202] [Q&A] [answered]
Hey there,

I am pretty sure I am missing something obvious here... is there a possibility to aggregate my data e.g. by day?

Simplified example: from JSON data like

```json
[
{"id":125,"status":"failure","started_at":"2021-11-18T10:28:48.861+01:00"},
{"id":124,"status":"success","started_at":"2021-11-18T09:28:48.861+01:00"},
{"id":123,"status":"success","started_at":"2021-11-17T10:28:48.861+01:00"}
]
```

I'd like to graph the following data:

```
| started-at | c |
| ---------- | - |
| 2021-11-18 | 2 |
| 2021-11-17 | 1 |
```

I know that I can use the "group by" transformation in order to count by a field's value, but in order to use it with `started_at` I would need a possibility to transform/truncate the timestamp to a date first, otherwise I just get `1` for each timestamp.
Unfortunately, I do not have control over the JSON datasource.

I do have a bit of experience with Grafana dashboards, but am currently taking my first steps with the Infinity Datasource. So maybe my approach is completely wrong, or maybe I just haven't found the clue yet.

Thanks a lot in advance for any ideas

**Comment by yesoreyeram** [comment, marked as answer]
Hi @jklingen - AFAIK, you can't truncate the date field neither in Grafana nor infinity. 

But expect something in near future ;) ( Before new year )

Thanks

**Reply by jklingen** [reply]
Hi @yesoreyeram, thanks for your quick response. Sounds promising 🤩

**Reply by deltamelter** [reply]
Hi  @yesoreyeram any news on this?

**Reply by yesoreyeram** [reply]
it is on track. I couldn't think of any easy way achieving this without messing the UI.. So decided to achieve this via "UQL".  Though it looks complicated, this will be more powerful in terms of customising the results. You can see a preview of UQL queries in https://grafana-infinity-datasource.herokuapp.com/d/uql 

PS: At the time of writing, the UQL preview dashboard doesn't have support for `startofday` function. But this will be available in 0.8.0-beta.2

![image](https://user-images.githubusercontent.com/153843/147593218-6d60b6c4-c54f-447d-922d-2bea5077a68f.png)

**Comment by cmwright-wcrclocal** [comment]
You can do this in Grafana 8.*, if you import the JSON started_at as a string, then use the Transform to change it to a Time type and use the date format YYYY-MM-DD. It should mark all the times as 00:00:00 and allow you to aggregate them by day.
![image](https://user-images.githubusercontent.com/74187695/148092156-74f60612-48f5-4906-940a-6c340d98442c.png)

**Reply by acasanova99** [reply]
Doesn't work in grafana 9.x

**Comment by deltamelter** [comment]
What other formats are available?  dow perhaps?

**Reply by yesoreyeram** [reply]
Does the link helpful? https://sriramajeyam.com/grafana-infinity-datasource/wiki/uql

**Reply by yesoreyeram** [reply]
There are several other functions like order by, limit also available.

**Comment by Usama420-git** [comment]
I have an date like `inviteDateTime` in this format `2022-11-01T11:03:06.68Z` how to group it by successfulCalls by day?

![asw](https://user-images.githubusercontent.com/51694624/209111924-bcabbc3d-166c-408e-8ec3-7e0a124f1e9b.PNG)

**Reply by yesoreyeram** [reply]
[refer the UQL query in screenshot](https://github.com/yesoreyeram/grafana-infinity-datasource/discussions/202#discussioncomment-1880113)

**Reply by Usama420-git** [reply]
I am using this but dont know how to get successfulCalls here I want to calculate true and false by day?

`parse-json
| scope "value"
| extend "inviteDateTime"=todatetime("inviteDateTime")
| extend "inviteDateTime"=startofday("inviteDateTime")
| summarize count(),"total"=mean("successfulCall") by "inviteDateTime"
| extend "inviteDateTime"=todatetime("inviteDateTime")`

**Reply by Usama420-git** [reply]
tried it like this also but getting error

![asa](https://user-images.githubusercontent.com/51694624/209144105-1e2a96d2-3203-4bc4-82fb-01ce5e0b8843.PNG)
![qq](https://user-images.githubusercontent.com/51694624/209144291-246feac1-5127-45cd-bc7a-02dc0fe1d159.PNG)

**Reply by yesoreyeram** [reply]
you need to wrap the field name "successfulCall" with double quote `"`. Also you can use jsonate inside your uql.

```kusto
parse-json 
| jsonata "value[successfulCall=true]"
| extend "inviteDateTime"=todatetime("inviteDateTime") 
| extend "inviteDateTime"=startofday("inviteDateTime")
| summarize count() by "inviteDateTime" , "successfulCall"
| extend "inviteDateTime"=todatetime("inviteDateTime")
```
